// @testing-library/react-native@14.0.0/dist/pure.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "act", {
  enumerable: true,
  get: function () {
    return _act.act;
  }
});
Object.defineProperty(exports, "cleanup", {
  enumerable: true,
  get: function () {
    return _cleanup.cleanup;
  }
});
Object.defineProperty(exports, "configure", {
  enumerable: true,
  get: function () {
    return _config.configure;
  }
});
Object.defineProperty(exports, "fireEvent", {
  enumerable: true,
  get: function () {
    return _fireEvent.fireEvent;
  }
});
Object.defineProperty(exports, "getDefaultNormalizer", {
  enumerable: true,
  get: function () {
    return _matches.getDefaultNormalizer;
  }
});
Object.defineProperty(exports, "isHiddenFromAccessibility", {
  enumerable: true,
  get: function () {
    return _accessibility.isHiddenFromAccessibility;
  }
});
Object.defineProperty(exports, "isInaccessible", {
  enumerable: true,
  get: function () {
    return _accessibility.isInaccessible;
  }
});
Object.defineProperty(exports, "render", {
  enumerable: true,
  get: function () {
    return _render.render;
  }
});
Object.defineProperty(exports, "renderHook", {
  enumerable: true,
  get: function () {
    return _renderHook.renderHook;
  }
});
Object.defineProperty(exports, "resetToDefaults", {
  enumerable: true,
  get: function () {
    return _config.resetToDefaults;
  }
});
Object.defineProperty(exports, "screen", {
  enumerable: true,
  get: function () {
    return _screen.screen;
  }
});
Object.defineProperty(exports, "userEvent", {
  enumerable: true,
  get: function () {
    return _userEvent.userEvent;
  }
});
Object.defineProperty(exports, "waitFor", {
  enumerable: true,
  get: function () {
    return _waitFor.waitFor;
  }
});
Object.defineProperty(exports, "waitForElementToBeRemoved", {
  enumerable: true,
  get: function () {
    return _waitForElementToBeRemoved.waitForElementToBeRemoved;
  }
});
Object.defineProperty(exports, "within", {
  enumerable: true,
  get: function () {
    return _within.within;
  }
});
var _act = require("./act");
var _cleanup = require("./cleanup");
var _fireEvent = require("./fire-event");
var _render = require("./render");
var _waitFor = require("./wait-for");
var _waitForElementToBeRemoved = require("./wait-for-element-to-be-removed");
var _within = require("./within");
var _config = require("./config");
var _accessibility = require("./helpers/accessibility");
var _matches = require("./matches");
var _renderHook = require("./render-hook");
var _screen = require("./screen");
var _userEvent = require("./user-event");
